### PR TITLE
opcode: add SHR opcode

### PR DESCRIPTION
**Pull request by SwordArt0820**
1. add SHR opcode
2. increase max degree of evm circuit from 9 to 11 due to 0..8 lagrange basis polynomial in SHR implementation. Will this be OK or any suggestion to optimize constraint degree in SHR?
3. Another two fixed lookup tables are added

### CODE REVIEW

**Comment by lispc** [comment]
don't change the permission of run.sh...

**Review by lispc — APPROVED** [review]

File: zkevm-circuits/src/evm_circuit/util/math_gadget.rs [hunk 1]
```diff
@@ -816,3 +817,349 @@ impl<F: FieldExt, const NUM_BYTES: usize> MaxGadget<F, NUM_BYTES> {
...
+                for idx in (8 - digit_transplacement)..8 {
+                    let nowidx = (virtual_idx * 8 + idx) as usize;
+                    equal_to_zero = equal_to_zero
+                        + (select_transplacement_polynomial.clone()
+                            * a_slice_front[nowidx].expr());
+                }
+            }
+        }
+        // for i in 1..32
+        // check shift[i] == 0
```
Comment (icemelon) [inline zkevm-circuits/src/evm_circuit/util/math_gadget.rs:975]: This is not correct. The shift can be greater than 256. Just the result will become 0 if shift > 256.

**Review by icemelon — COMMENTED** [review]
@SwordArt0820 High-level comments after the first look
- You should first write down the circuit spec in a markdown doc before we review the circuit code
- need more comments and docs in the code
- better variable naming

File: zkevm-circuits/src/evm_circuit/util/math_gadget.rs [hunk 2]
```diff
@@ -816,3 +817,349 @@ impl<F: FieldExt, const NUM_BYTES: usize> MaxGadget<F, NUM_BYTES> {
         Ok(select::value(lt, rhs, lhs))
     }
 }
+
+//generating Lagrange base polynomial for a cell.
+//the polynomial will equal to 1 when cell equals to idx, otherwise 0.
+//the cell's domain will be 0..domain_size
```
[the remark this answers is not in the document]
Comment (SwordArt0820) [reply zkevm-circuits/src/evm_circuit/util/math_gadget.rs:823]: > If there's a constraint on the value of cell, do you also need to add extra constraint on that as well?

The range of every cell which uses this function to generate the polynomial will be check by a lookup below.So there is no need to add extra constraint.

**Comment by icemelon** [comment]
Need to close and reopen a new PR due to some bug in Github that causes changes from `main` to appear in the PR

**Pull request closed without merging**